lib/xcode/install.rb

```diff
@@ -245,20 +246,135 @@ def xcode_license_approved?
+    end
+
+    def installed?
+      # FIXME: use downloadables' `InstalledIfAllReceiptsArePresentOrNewer` key
+      File.directory?(@install_prefix)
+    end
+
+    def installed_string
+      installed? ? 'installed' : 'not installed'
+    end
+
+    def to_s
+      "#{name} (#{installed_string})"
+    end
+
+    def xcode
+      Installer.new.installed_versions.find do |x|
+        x.available_simulators.find do |s|
+          s.version == version
+        end
+      end
+    end
+
+    def download
+      result = Curl.new.fetch(source, CACHE_DIR)
+      result ? dmg_path : nil
+    end
+
+    def install
+      download unless dmg_path.exist?
+      prepare_package unless pkg_path.exist?
+      puts "Please authenticate to install #{name}..."
+      `sudo installer -pkg #{pkg_path} -target /`
+      fail Informative, "Could not install #{name}, please try again" unless installed?
+      puts "Successfully installed #{name}"
+    end
+
+    :private
+
+    def prepare_package
+      puts 'Mounting DMG'
+      mount_location = `hdiutil mount -nobrowse -noverify #{dmg_path}`.scan(/\/Volumes.*\n/).first.chomp
+      puts 'Expanding pkg'
+      expanded_pkg_path = CACHE_DIR + identifier
+      FileUtils.rm_rf(expanded_pkg_path)
+      `pkgutil --expand #{mount_location}/*.pkg #{expanded_pkg_path}`
+      puts "Expanded pkg into #{expanded_pkg_path}"
+      puts 'Unmounting DMG'
+      `umount #{mount_location}`
+      puts 'Setting package installation location'
+      package_info_path = expanded_pkg_path + 'PackageInfo'
+      package_info_contents = File.read(package_info_path)
+      File.open(package_info_path, 'w') do |f|
+        f << package_info_contents.sub('pkg-info', %(pkg-info install-location="#{@install_prefix}"))
+      end
+      puts 'Rebuilding package'
+      `pkgutil --flatten #{expanded_pkg_path} #{pkg_path}`
+      FileUtils.rm_rf(expanded_pkg_path)
+    end
+
+    def dmg_path
+      CACHE_DIR + Pathname.new(source).basename
+    end
+
+    def pkg_path
+      CACHE_DIR + "#{identifier}.pkg"
+    end
+
+    def apply_variables(template)
+      variable_map = {
+        '$(DOWNLOADABLE_VERSION_MAJOR)' => version.to_s.split('.')[0],
+        '$(DOWNLOADABLE_VERSION_MINOR)' => version.to_s.split('.')[1],
+        '$(DOWNLOADABLE_IDENTIFIER)' => identifier,
+        '$(DOWNLOADABLE_VERSION)' => version.to_s
+      }.freeze
+      variable_map.each do |key, value|
+        next unless template.include?(key)
+        template.sub!(key, value)
+      end
+      template
+    end
+  end
+
   class InstalledXcode
     attr_reader :path
     attr_reader :version
+    attr_reader :uuid
+    attr_reader :downloadable_index_url
+    attr_reader :available_simulators
 
     def initialize(path)
       @path = Pathname.new(path)
-      @version = get_version(path)
+    end
+
+    def version
+      @version ||= Gem::Version.new(plist_entry(':DTXcode').to_i.to_s.split(//).join('.'))
```

Comment:
This is a new way to get the version and it's a change from what's there now (will print `7.0.0` rather than `7.0`). I'll happily split this into a new `full_version` property if you prefer to maintain the old one too.